streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/RackUtils.java

```diff
@@ -0,0 +1,131 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RackUtils {
+    private static final Logger LOG = LoggerFactory.getLogger(RackUtils.class);
+
+    public static Map<TopicPartition, Set<String>> getRacksForTopicPartition(final Cluster cluster,
+                                                                             final InternalTopicManager internalTopicManager,
+                                                                             final Set<TopicPartition> topicPartitions,
+                                                                             final boolean isChangelog) {
+        final Set<String> topicsToDescribe = new HashSet<>();
+        if (isChangelog) {
+            topicsToDescribe.addAll(topicPartitions.stream().map(TopicPartition::topic).collect(
+                Collectors.toSet()));
+        } else {
+            topicsToDescribe.addAll(topicsWithStaleMetadata(cluster, topicPartitions));
+        }
+
+        final Set<TopicPartition> topicsWithUpToDateMetadata = topicPartitions.stream()
+            .filter(partition -> !topicsToDescribe.contains(partition.topic()))
+            .collect(Collectors.toSet());
+        final Map<TopicPartition, Set<String>> racksForTopicPartition = knownRacksForPartition(
+            cluster, topicsWithUpToDateMetadata);
+
+        final Map<String, List<TopicPartitionInfo>> freshTopicPartitionInfo =
+            describeTopics(internalTopicManager, topicsToDescribe);
+        freshTopicPartitionInfo.forEach((topic, partitionInfos) -> {
+            partitionInfos.forEach(partitionInfo -> {
+                final int partition = partitionInfo.partition();
+                final TopicPartition topicPartition = new TopicPartition(topic, partition);
+                final List<Node> replicas = partitionInfo.replicas();
+                if (replicas == null || replicas.isEmpty()) {
+                    LOG.error("No replicas found for topic partition {}: {}", topic, partition);
+                    return;
+                }
+
+                final Set<String> racks = replicas.stream().filter(Node::hasRack).map(Node::rack).collect(
+                    Collectors.toSet());
+                racksForTopicPartition.computeIfAbsent(topicPartition, k -> new HashSet<>());
+                racksForTopicPartition.get(topicPartition).addAll(racks);
+            });
+        });
+
+        return racksForTopicPartition;
+    }
+
+    public static Set<String> topicsWithStaleMetadata(final Cluster cluster, final Set<TopicPartition> topicPartitions) {
+        final Set<String> topicsWithStaleMetadata = new HashSet<>();
+        for (final TopicPartition topicPartition : topicPartitions) {
+            final PartitionInfo partitionInfo = cluster.partition(topicPartition);
+            if (partitionInfo == null) {
+                LOG.error("TopicPartition {} doesn't exist in cluster", topicPartition);
+                continue;
```

Comment:
It's not an error for a topic to not be included in the `Cluster`, even source topics might not exist here if they had to be created by the assignor during the rebalance, since the `Cluster` metadata represents the state of the cluster when this rebalance/assignment first began.

Since the point of this method seems to be to collect topics with missing metadata that we'll need to look up via a DescribeTopics request, the ones for which `cluster.partition(topicPartition)` returns `null` are exactly the ones that should be returned by this method.

In fact I'd go ahead and remove everything past this line as well, this method should focus only on differentiating topics with missing metadata from ones we already have the info for. If the `Cluster` has metadata for this partition but the `replicas` set is missing/empty, then there's something wrong with this partition, and calling DescribeTopics probably won't help

Let's rename this to `#topicsWithMissingMetadata` while we're at it

Reply:
This isn't how the current rack aware code works:
```            for (final TopicPartition topicPartition : topicPartitions) {
                final PartitionInfo partitionInfo = fullMetadata.partition(topicPartition);
                if (partitionInfo == null) {
                    log.error("TopicPartition {} doesn't exist in cluster", topicPartition);
                    return false;
                }
                final Node[] replica = partitionInfo.replicas();
                if (replica == null || replica.length == 0) {
                    topicsToDescribe.add(topicPartition.topic());
                    continue;
                }
                for (final Node node : replica) {
                    if (node.hasRack()) {
                        racksForPartition.computeIfAbsent(topicPartition, k -> new HashSet<>()).add(node.rack());
                    } else {
                        log.warn("Node {} for topic partition {} doesn't have rack", node, topicPartition);
                        return false;
                    }
                }
            }
```
Above is the logic of `populateTopicsToDescribe`, which evidently uses the `(replica == null || replica.length == 0)` condition to decide to fetch further topic information, yet considers `cluster.partition(topicPartition)` to be an error.